src/Native/Runtime/arm64/WriteBarriers.S

```diff
@@ -1,33 +1,368 @@
+        //
+        // On exit:
+        //   destReg:   trashed
+        //
+        .macro INSERT_CHECKED_WRITE_BARRIER_CORE destReg, refReg, trash, trash2
+
+        // The "check" of this checked write barrier - is destReg
+        // within the heap? if no, early out.
+        adrp    x\trash, g_lowest_address
+        add	    x\trash, x\trash, :lo12:g_lowest_address
+        ldr     x\trash, [x\trash]
+        cmp     \destReg, x\trash
+        blt     0f
+
+        adrp    x\trash, g_highest_address
+        add	    x\trash, x\trash, :lo12:g_highest_address
+        ldr     x\trash, [x\trash]
+        cmp     \destReg, x\trash
+        bgt     0f
+
+        INSERT_UNCHECKED_WRITE_BARRIER_CORE \destReg, \refReg, \trash, \trash2
+
+0:
+        // Exit label
+    .endm
+
+// RhpCheckedAssignRef(Object** dst, Object* src)
 //
-// RhpByRefAssignRef simulates movs instruction for object references.
+// Write barrier for writes to objects that may reside
+// on the managed heap.
 //
 // On entry:
-//  x0: address of ref-field (assigned to)
-//  x1: address of the data (source)
-//  x3: be trashed
+//   x0 : the destination address (LHS of the assignment).
+//        May not be an object reference (hence the checked).
+//   x1 : the object reference (RHS of the assignment).
+// On exit:
+//   x1 : trashed
+//   x9 : trashed
+    LEAF_ENTRY RhpCheckedAssignRef, _TEXT
+    ALTERNATE_ENTRY RhpCheckedAssignRefAVLocation
+    ALTERNATE_ENTRY RhpCheckedAssignRefX1
+    ALTERNATE_ENTRY RhpCheckedAssignRefX1AVLocation
+
+        stlr    x1, [x0]
+        
+        INSERT_CHECKED_WRITE_BARRIER_CORE x0, x1, 9, x1
+        
+        ret
+
+    LEAF_END RhpCheckedAssignRef, _TEXT
+
+// RhpAssignRef(Object** dst, Object* src)
 //
+// Write barrier for writes to objects that are known to
+// reside on the managed heap.
+//
+// On entry:
+//  x0 : the destination address (LHS of the assignment).
+//  x1 : the object reference (RHS of the assignment).
 // On exit:
-//  x0, x1 are incremented by 8, 
-//  x3: trashed
+//  x1 : trashed
+//  x9 : trashed
+    LEAF_ENTRY RhpAssignRef, _TEXT
+    ALTERNATE_ENTRY RhpAssignRefAVLocation
+    ALTERNATE_ENTRY RhpAssignRefX1
+    ALTERNATE_ENTRY RhpAssignRefX1AVLocation
+
+        stlr    x1, [x0]
+ 
+        INSERT_UNCHECKED_WRITE_BARRIER_CORE x0, x1, 9, x1
+ 
+        ret
+
+    LEAF_END RhpAssignRef, _TEXT
+
+// Interlocked operation helpers where the location is an objectref, thus requiring a GC write barrier upon
+// successful updates. 
+
+// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
+// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpCheckedLockCmpXchgAVLocation
+// - Function "UnwindWriteBarrierToCaller" assumes no registers where pushed and LR contains the return address
+
+// RhpCheckedLockCmpXchg(Object** dest, Object* value, Object* comparand)
+//
+// Interlocked compare exchange on objectref.
```

Comment:
The asm code here and in Interlocked.S (that is preexisting this PR) is missing memory barrier. The interlocked operations in .NET are expected to issue a membarrier even when the operation fails. 

(https://github.com/dotnet/runtime/blob/6ca600396ee2e80e8bf93c72c2b5439aadc86073/src/coreclr/src/pal/inc/pal.h#L3210 in CoreCLR talks about it.)

Reply:
I might be overlooking something here but WriteBarrier.asm uses the same code. For references I checked the code that the compiler generates when I am using __sync_val_compare_and_swap. It's the same code, too. 

Reply:
WriteBarrier.asm has the same issue. It would be nice to fix it there too.

The gold standard is what CoreCLR does in these cases. CoreCLR issues full barrier here, so CoreRT should match it. The CoreCLR implementation of `RhpCheckedLockCmpXchg` is in C++ https://github.com/dotnet/runtime/blob/29e9b5b7fd95231d9cd9d3ae351404e63cbb6d5a/src/coreclr/src/vm/comutilnative.cpp#L1669 . It gets the full barrier via `PAL_ArmInterlockedOperationBarrier` (https://github.com/dotnet/runtime/blob/6ca600396ee2e80e8bf93c72c2b5439aadc86073/src/coreclr/src/pal/inc/pal.h#L3210).

Reply:
I think I got it. So I will just add a new ArmInterlockedOperationBarrier macro that contains a "dmb ish" to make sure the assembler function behaves in the same way as the C function in the regular runtime. At least my LLVM based compiler does this for __sync_synchronize() which is used by PAL_ArmInterlockedOperationBarrier

Reply:
Yes, you got it. You can add a macro; or just add `dmb ish` directly in the few places that need it. Either way is fine with me.